jsparagus/gen.py

```diff
@@ -881,14 +882,26 @@ def __init__(self, index, locations, delayed_actions=OrderedFrozenSet()):
         # parse-table, as well as the actions which have not yet been executed.
         def hashed_content():
             for item in sorted(self.locations):
-                yield item.prod_index
-                yield item.offset
+                yield item
+                yield "\n"
             yield "delayed_actions"
-            for action in sorted(delayed_actions):
+            for action in self.delayed_actions:
                 yield hash(action)
 
         self._hash = hash(tuple(hashed_content()))
 
+    def stable_hash(self, state_hash):
+        def hashed_content():
```

Comment:
With this change, we'll have the same information—what defines a state—in three places:
- the `_hash` computation above,
- `stable_hash`
- `__eq__`

It's a small amount of code that's repeated, so I'm not super worried.

Still, please consider computing this hash in the constructor and storing it in a property `self.stable_hash`, because then we can delete `ParseTable.state_hash`. Where `state_hash` is passed around now, we can instead pass around `states` and use `states[i].stable_hash`.

Reply:
Sounds good!, this will also avoid having to maintain the `state_hash` list sorted the same way each time we rewrite the `states` list.